### PR TITLE
Stack Recipe Lists

### PR DESCRIPTION
**Pull request by ghost**
Ports
Baystation12/Baystation12#3143
And some later changes

Basically, when you use metal in hand to bring up the recipes, similar recipes can be put into a stack_recipe_list, so the recipe menu is not spammed with similar recipes. You click the recipe (example: "office chairs", and that brings up another menu with more recipes "dark office chair (5 sheets)" to make

stacklist1
stacklist2

Also made stack menus use /datum/browser

:cl:
add: You can make many different types of office and comfy chairs with metal
tweak: Stack menus use /datum/browser
/:cl:

### CODE REVIEW

File: code/game/objects/items/stacks/stack.dm [hunk 1]
```diff
@@ -86,60 +87,78 @@
...
+				for (var/n in multipliers)
+					if (max_multiplier>=n)
+						t1 += " <A href='?src=\ref[src];make=[i];multiplier=[n]'>[n*R.res_amount]x</A>"
+				if (!(max_multiplier in multipliers))
+					t1 += " <A href='?src=\ref[src];make=[i];multiplier=[max_multiplier]'>[max_multiplier*R.res_amount]x</A>"
 
 	t1 += "</TT></body></HTML>"
 	user << browse(t1, "window=stack")
 	onclose(user, "stack")
+	return
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/stack.dm]: no need for this trailing return

File: code/game/objects/items/stacks/stack.dm [hunk 2]
```diff
@@ -189,6 +208,8 @@
 
 	if (src && usr.machine==src) //do not reopen closed window
 		addtimer(CALLBACK(src, /atom/.proc/interact, usr), 0)
+		return
+	return
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/stack.dm]: same as above for these two

File: code/game/objects/items/stacks/stack.dm [hunk 3]
```diff
@@ -86,60 +87,78 @@
...
-	var/t1 = text("<HTML><HEAD><title>Constructions from []</title></HEAD><body><TT>Amount Left: []<br>", src, get_amount())
-	for(var/i=1;i<=recipes.len,i++)
-		var/datum/stack_recipe/R = recipes[i]
-		if (isnull(R))
+	var/list/recipe_list = recipes
+	if (recipes_sublist && recipe_list[recipes_sublist] && istype(recipe_list[recipes_sublist], /datum/stack_recipe_list))
+		var/datum/stack_recipe_list/srl = recipe_list[recipes_sublist]
+		recipe_list = srl.recipes
+	var/t1 = text("<HTML><HEAD><title>Constructions from []</title></HEAD><body><TT>Amount Left: []<br>", src, src.get_amount())
+	for(var/i=1;i<=recipe_list.len,i++)
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/stack.dm]: `for(var/i in 1 to length(recipe_list))`

File: code/game/objects/items/stacks/stack.dm [hunk 4]
```diff
@@ -86,60 +87,78 @@
 		. = (amount)
 
 /obj/item/stack/attack_self(mob/user)
-	interact(user)
+	list_recipes(user)
 
-/obj/item/stack/interact(mob/user)
+/obj/item/stack/proc/list_recipes(mob/user, recipes_sublist)
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/stack.dm]: use interact, not a new proc

File: code/game/objects/items/stacks/stack.dm [hunk 5]
```diff
@@ -345,3 +366,14 @@
 	src.one_per_turf = one_per_turf
 	src.on_floor = on_floor
 	src.window_checks = window_checks
+
+/*
+ * Recipe list datum
+ */
+/datum/stack_recipe_list
+	var/title = "ERROR"
+	var/list/recipes = null
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/stack.dm]: unnecessary null

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by fludd12** [comment]
Can we get their fancy colored comfy chairs now?

**Pull request merged**